File: ruby/trie.rb
# Reference https://en.wikipedia.org/wiki/Trie
# Example
# t = Trie.new
# t.insert("hello")
# t.search("hell") => false
# t.starts_with("hell") => true
class Trie
  attr_accessor :children, :is_word

  def initialize()
    @children = {}
    @is_word = false
  end

  def insert(word)
    children = @children
    word.each_char.each_with_index do |char, index|
      children[char] ||= Trie.new
      children[char].is_word = true if word.size == index + 1
      children = children[char].children
    end
  end

  # Returns true if the word is in the trie.
  def search(word)
    match_helper(word, true)
  end

  # Returns true if there is any word in the trie that starts with the given prefix.
  def starts_with(prefix)
    match_helper(prefix)
  end

  private

  def match_helper(match, is_word = false)
    children = @children
    found = false
    match.each_char.each_with_index do |char, index|
      break if children[char].nil?
      if match.size == index + 1 && requires_word?(is_word, children[char].is_word)
        found = true
      end
      children = children[char].children
    end
    found
  end

  def requires_word?(requires_word, node_is_word)
    return true unless requires_word
    node_is_word
  end
end

File: golang/trie.go
package trie

import (
  "fmt"
)

type Trie struct {
  children map[string]*Trie
  isWord bool
}

func NewTrie() *Trie {
  return &Trie{isWord: false, children: make(map[string]*Trie)}

}

func (t *Trie) Insert(word string) {
  children := t.children
  for index, char := range word {
    if(children[string(char)] == nil) {
      children[string(char)] = &Trie{isWord: false, children: make(map[string]*Trie)}
    }

    if(len(word) == index + 1) {
      children[string(char)].isWord = true
    }
    children = children[string(char)].children
  }
}


func (t *Trie) Search(word string) (found bool) {
  return t.searchHelper(word, true)
}

func (t *Trie) StartsWith(word string) (found bool) {
  return t.searchHelper(word, false)
}

func (t *Trie) searchHelper(word string, wordRequired bool) (found bool) {
  children := t.children
  found = false
  for index, char := range word {
    if(children[string(char)] == nil) {
      break
    } else {
      if(len(word) == index + 1 && requiresWord(wordRequired,children[string(char)].isWord)) {
        found = true
      }
    }
    children = children[string(char)].children
  }
  return found
}

func requiresWord(condition bool, isWord bool) bool {
  if(condition == false) {
    return true
  }
  return condition && isWord
}
